File: day18/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatalln(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatalln(err)
		}
	}()

	cloud := map[Point]struct{}{}
	first := true
	var start Point

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		strcoords := strings.Split(scanner.Text(), ",")
		if len(strcoords) != 3 {
			log.Fatalln("Invalid line")
		}
		x, err := strconv.Atoi(strcoords[0])
		if err != nil {
			log.Fatalln(err)
		}
		y, err := strconv.Atoi(strcoords[1])
		if err != nil {
			log.Fatalln(err)
		}
		z, err := strconv.Atoi(strcoords[2])
		if err != nil {
			log.Fatalln(err)
		}
		cloud[Point{x: x, y: y, z: z}] = struct{}{}
		if first {
			first = false
			start = Point{x: x, y: y, z: z}
		}
	}

	if err := scanner.Err(); err != nil {
		log.Fatalln(err)
	}

	border := map[Point]struct{}{}

	surfaceArea := 0
	for k := range cloud {
		surfaceArea += 6 - numNeighbors(k, cloud)
		for _, i := range interCardinalDirs {
			p := k.add(i)
			if _, ok := cloud[p]; !ok {
				border[p] = struct{}{}
				if p.x > start.x {
					start = p
				}
			}
		}
	}
	fmt.Println("Part 1:", surfaceArea)

	extSurfaceArea := 0
	openSet := []Point{start}
	closedSet := map[Point]struct{}{start: {}}
	for len(openSet) != 0 {
		l := len(openSet) - 1
		p := openSet[l]
		openSet = openSet[:l]

		for _, i := range cardinalDirs {
			k := p.add(i)
			if _, ok := cloud[k]; ok {
				extSurfaceArea++
			} else if _, ok := border[k]; ok {
				if _, ok := closedSet[k]; !ok {
					openSet = append(openSet, k)
					closedSet[k] = struct{}{}
				}
			}
		}
	}
	fmt.Println("Part 2:", extSurfaceArea)
}

type (
	Point struct {
		x, y, z int
	}
)

func (p Point) add(other Point) Point {
	return Point{
		x: p.x + other.x,
		y: p.y + other.y,
		z: p.z + other.z,
	}
}

var (
	cardinalDirs = []Point{
		{1, 0, 0},
		{-1, 0, 0},
		{0, 1, 0},
		{0, -1, 0},
		{0, 0, 1},
		{0, 0, -1},
	}
	interCardinalDirs = []Point{
		{-1, -1, -1},
		{-1, -1, 0},
		{-1, -1, 1},
		{-1, 0, -1},
		{-1, 0, 0},
		{-1, 0, 1},
		{-1, 1, -1},
		{-1, 1, 0},
		{-1, 1, 1},
		{0, -1, -1},
		{0, -1, 0},
		{0, -1, 1},
		{0, 0, -1},
		{0, 0, 1},
		{0, 1, -1},
		{0, 1, 0},
		{0, 1, 1},
		{1, -1, -1},
		{1, -1, 0},
		{1, -1, 1},
		{1, 0, -1},
		{1, 0, 0},
		{1, 0, 1},
		{1, 1, -1},
		{1, 1, 0},
		{1, 1, 1},
	}
)

func numNeighbors(p Point, cloud map[Point]struct{}) int {
	count := 0
	for _, i := range cardinalDirs {
		if _, ok := cloud[p.add(i)]; ok {
			count++
		}
	}
	return count
}

File: day18/src/main.rs
use std::collections::HashSet;
use std::fs::File;
use std::hash::Hash;
use std::io::prelude::*;
use std::io::BufReader;
use std::ops::Add;

const PUZZLEINPUT: &str = "input.txt";

type BoxError = Box<dyn std::error::Error>;
type BoxResult<T> = Result<T, BoxError>;

fn main() -> BoxResult<()> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let mut cloud = HashSet::new();
    let mut start = None;

    for line in reader.lines() {
        let line = line?;
        let p = if let [x, y, z] = line
            .split(",")
            .flat_map(|s| s.parse::<i32>())
            .collect::<Vec<_>>()[..]
        {
            Point::new(x, y, z)
        } else {
            return Err("Invalid line".into());
        };
        cloud.insert(p);
        if start == None {
            start = Some(p);
        }
    }

    let mut start = if let Some(v) = start {
        v
    } else {
        return Err("No points".into());
    };

    let mut border = HashSet::new();

    let mut surface_area = 0;
    for &k in &cloud {
        surface_area += 6 - CARDINAL_DIRS
            .iter()
            .filter(|&&i| cloud.contains(&(k + i)))
            .count();
        for &i in INTER_CARDINAL_DIRS {
            let p = k + i;
            if !cloud.contains(&p) {
                border.insert(p);
                if p.x > start.x {
                    start = p;
                }
            }
        }
    }
    println!("Part 1: {}", surface_area);

    let mut ext_surface_area = 0;
    let mut open_set = vec![start];
    let mut closed_set = HashSet::new();
    closed_set.insert(start);
    while let Some(p) = open_set.pop() {
        for &i in CARDINAL_DIRS {
            let k = p.add(i);
            if cloud.contains(&k) {
                ext_surface_area += 1;
            } else if border.contains(&k) && !closed_set.contains(&k) {
                open_set.push(k);
                closed_set.insert(k);
            }
        }
    }
    println!("Part 2: {}", ext_surface_area);

    Ok(())
}

#[derive(Clone, Copy, PartialEq, Eq, Hash)]
struct Point {
    x: i32,
    y: i32,
    z: i32,
}

impl Point {
    const fn new(x: i32, y: i32, z: i32) -> Self {
        Point { x, y, z }
    }
}

impl Add for Point {
    type Output = Self;

    fn add(self, other: Self) -> Self {
        Self {
            x: self.x + other.x,
            y: self.y + other.y,
            z: self.z + other.z,
        }
    }
}

const CARDINAL_DIRS: &[Point] = &[
    Point::new(1, 0, 0),
    Point::new(-1, 0, 0),
    Point::new(0, 1, 0),
    Point::new(0, -1, 0),
    Point::new(0, 0, 1),
    Point::new(0, 0, -1),
];

const INTER_CARDINAL_DIRS: &[Point] = &[
    Point::new(-1, -1, -1),
    Point::new(-1, -1, 0),
    Point::new(-1, -1, 1),
    Point::new(-1, 0, -1),
    Point::new(-1, 0, 0),
    Point::new(-1, 0, 1),
    Point::new(-1, 1, -1),
    Point::new(-1, 1, 0),
    Point::new(-1, 1, 1),
    Point::new(0, -1, -1),
    Point::new(0, -1, 0),
    Point::new(0, -1, 1),
    Point::new(0, 0, -1),
    Point::new(0, 0, 1),
    Point::new(0, 1, -1),
    Point::new(0, 1, 0),
    Point::new(0, 1, 1),
    Point::new(1, -1, -1),
    Point::new(1, -1, 0),
    Point::new(1, -1, 1),
    Point::new(1, 0, -1),
    Point::new(1, 0, 0),
    Point::new(1, 0, 1),
    Point::new(1, 1, -1),
    Point::new(1, 1, 0),
    Point::new(1, 1, 1),
];
